File: boilerplate/nestjs/order-service/src/application/usecases/place-order.use-case.impl.ts
// application/usecases/place-order.use-case.impl.ts
import { Injectable } from '@nestjs/common';
import { PlaceOrderUseCase } from './place-order.use-case.interface';
import { PlaceOrderDto } from '../dtos/place-order.dto';
import { OrderResponseDto } from '../dtos/order-response.dto';
import { Order } from '@domain/models/order.aggregate';
import { OrderId } from '@domain/models/order-id.value-object';
import { OrderItem } from '@domain/models/order-item.value-object';
import { OrderRepositoryPort } from '@domain/ports/order-repository.port';
import { Decimal } from 'decimal.js';
import { v4 as uuidv4 } from 'uuid';

@Injectable()
export class PlaceOrderUseCaseImpl implements PlaceOrderUseCase {
  constructor(private readonly orderRepository: OrderRepositoryPort) {}

  async execute(dto: PlaceOrderDto): Promise<OrderResponseDto> {
    const items = dto.items.map(
      (item) =>
        new OrderItem(
          item.productId,
          item.quantity,
          new Decimal(item.unitPrice),
        ),
    );

    const order = new Order(new OrderId(uuidv4()), items);
    const confirmedOrder = order.confirm();

    await this.orderRepository.save(confirmedOrder);

    return {
      orderId: confirmedOrder.id.value,
      status: confirmedOrder.status,
      totalAmount: confirmedOrder.totalAmount().toString(),
      createdAt: confirmedOrder.createdAt,
    };
  }
}

File: tools/cookiecutter-python/{{cookiecutter.project_slug}}/order-service/src/application/usecases/place_order_use_case_impl.py
from __future__ import annotations

from decimal import Decimal
from typing import List
from uuid import UUID

from application.dtos import CreateOrderCommand, OrderResult
from domain.order import Order, OrderItem
from domain.services.order_placement_service import OrderPlacementService


class PlaceOrderUseCaseImpl:
    """Application use case: maps commands, orchestrates domain service, returns result.

    Follows CQS: this is a Command (mutates state, returns result DTO).
    """

    def __init__(self, placement_service: OrderPlacementService):
        self._placement_service = placement_service

    def execute(self, command: CreateOrderCommand) -> OrderResult:
        """Execute the place order use case.

        Steps:
        1. Map DTO → domain objects (Domain Item uses Decimal)
        2. Delegate to OrderPlacementService (domain)
        3. Map domain result → application DTO
        """
        items: List[OrderItem] = [
            OrderItem(
                product_id=item.product_id,
                quantity=item.quantity,
                unit_price=item.unit_price,
            )
            for item in command.items
        ]

        saved = self._placement_service.place_order(
            customer_id=command.customer_id,
            items=items,
        )

        return OrderResult(
            order_id=saved.id.value,
            status=saved.status,
            created_at=saved.created_at,
        )
